### DISCUSSION TITLE
How To Perform Massive Concurrent Transactions in ADempiere

### DISCUSSION THREAD

**Discussion by marcalwestf** [adempiere/adempiere#4104] [Show and tell]
# Motivation
Sometimes, it is necessary to test ADempiere massively. A massive test stresses the application at all levels (hardware, operating system, database, application server, processes, java classes) and exposes weaknesses.

This can be achieved on an ADempiere installation not only by creating a document and sequentially completing it, but by creating hundreds or even thousands of documents and completing them concurrently. 

High stress can be achieved if you create 2 clients and you import in parallel using different browsers.
Even more stress can be achieved if you use 2 machines, 2 clients each and do the import in parallel.

The following instructions can create and complete 12,000 documents or more per client. If for example, you run this tests simutaneously on 4 clients, there will be about 48,000 documents that ADempiere will try to execute at the same time.

**It is recommended to start with a small import and only when this small import is successful, move on to the massive tests**.

In the next sections, it will be explained how to create an environment to test an ADempiere installation massively.

## Structure of the tests
The tests will consist of the following steps:
- Product import
- Import products to _Price Lists_
- Import and execute _Purchase Orders_
- Create _Material Receipts_ out of _Purchase Orders_
- Create and execute _Purchase Invoices_
- Import and execute _Sales Orders_ that includes _Sales Invoices_ and _Material Issue_

The steps will include accounting processing where applicable.

## Prerequisites
All you need is a standard ADempiere installation and a functioning database.
It is recommendabe to have a clean, pristine database.
If you want to have all control, better do the tests on a new client.

## Massive Product Import
In this section, more than **10,000** products will be imported.

#### Step 1: the _Import Loader Format_ must be created (header and lines) as follows:
![01a_Importar_Productos-Encabezado](https://user-images.githubusercontent.com/1789408/236874540-ace46549-d578-48e8-99e1-efd823993983.png)
![01b_Importar_Productos-Lineas](https://user-images.githubusercontent.com/1789408/236874572-4eaa4889-4803-47a9-8e1c-a8468e6b5302.png)

In _AD_CLient_ID_ you can put the ID of your choice.
It is recommendable to create a new client for every test.
The more clients are created and tested, the heavier the test.

#### Step 2: the _Import File Loader_  must be executed.
Remember that the _Import Loader Format_ and the _Import File_ must coincide.
![01d_ImportFileLoader](https://user-images.githubusercontent.com/1789408/236876093-2af16d94-dcfb-4681-9f5e-4a59993a5326.png)

The _Import File_ :
[01c_Importar_Productos.csv](https://github.com/adempiere/adempiere/files/11422893/01c_Importar_Productos.csv)

## Create Price Lists
In this section, a Purchase and a _Sales Price List_ will be created. They will include all imported products.
#### Step 2: Import products into _Standard Price List_.
Create the _Import Loader Format_:
![02a_Importar_PriceList-Encabezado](https://user-images.githubusercontent.com/1789408/236877904-398916d5-218c-4018-8410-722e4098075a.png)
![02b_Importar_PriceList-Lineas](https://user-images.githubusercontent.com/1789408/236877924-7a0f510d-3977-4931-85d4-5bc58271e169.png)

Make sure the constants for _AD_Client_ID_, _M_PriceList_Version_ID_ and _M_PriceList_ID_ match the IDs of the Client, Price List Version and Price List you have created.

#### Step 3: the _Import File Loader_  must be executed.
Remember that the _Import Loader Format_ and the _Import File_ must coincide.
The _Import File_ :
[02c_Importar_PriceList.csv](https://github.com/adempiere/adempiere/files/11422984/02c_Importar_PriceList.csv)

#### Step 6: create a _Sales Price List_ and a _Sales Price List Version_ directly from the _Standard Price List_ created, using the _Price Schema_ that **increases** the amount.
Take note of all Price Lists and Price List Versions created.
They will be different depending on the database you are using and the quantity of clients involved in the test.

## Massive Import of Purchase Orders
Few Purchase Orders with hundreds of lines each.

#### Step 1: Create the _Import Loader Format_:
![03a_Importar_Ordenes_de_Compras-Encabezado](https://user-images.githubusercontent.com/1789408/236882008-a893f5a7-1668-4905-a385-e408c9056389.png)
![03b_Importar_Ordenes_de_Compras-Lineas](https://user-images.githubusercontent.com/1789408/236882021-3183f345-d93d-4b31-97af-257125a5f5d3.png)
Make sure the constants used in the  _Import Loader Format Lines_ are in sync with the values of the Price List, Document Type, Tax, etc. you are using.

#### Step 2: the _Import File Loader_  must be executed.
Remember that the _Import Loader Format_ and the _Import File_ must coincide.
The _Import Files_ :
[03c_Importar_Ordenes_de_Compras.csv](https://github.com/adempiere/adempiere/files/11423191/03c_Importar_Ordenes_de_Compras.csv)
[03d_Importar_Ordenes_de_Compras_Februar_2022.csv](https://github.com/adempiere/adempiere/files/11423442/03d_Importar_Ordenes_de_Compras_Februar_2022.csv)
[03e_Importar_Ordenes_de_Compras_Februar_2022_verkürzt.csv](https://github.com/adempiere/adempiere/files/11423443/03e_Importar_Ordenes_de_Compras_Februar_2022_verkurzt.csv)

Here, three files are delivered. You can use one or all of them.
When importing, **check to complete the Purchase Orders**.
There are some Purchase Orders with hundreds of lines each.

## Massive Import of Sales Invoices
This constitutes the actual massive test, because thousands of _On Credit Sales Orders_ are imported and executed. Some consist of a single line, some have many lines.

As a result, these  _On Credit Sales Orders_ create and complete their corresponding _Material Receipt_ (issue), _Sales Invoice_ and _Accounting_ (you must configure ADempiere to apply accounting immediately).

If we take for example the file with 3,000 Sales Orders consisting of one line per invoice, the result will be the creation and completion of 12,000 documents.

If several clients are involved, this will generate a huge CPU and process demand, stressing all resources.

#### Step 1: Create the _Import Loader Format_
Similar to former steps.
Important here are Business Partner, Price List, Warehouse, Document Type. All of them must be created manually before, with all data needed (e.g. Business Partner as a customer and having an address), their IDs written down and included in the files.

#### Step 2: the _Import File Loader_  must be executed.
Similar to former steps. Make sure the constants correspond to the values you use.
Remember that the _Import Loader Format_ and the _Import File_ must coincide.

Import Files: 
Here, several files are delivered. You can use one or all of them.
When importing, check to complete the _Sales Orders_.
The quantity of every line is 1 or 2. This makes sure that the existences don't get depleted.

#### Sales Orders consisting of hundred or less lines. Altogether 17,256 Lines : 
[03h_Importar_Ordenes_de_Ventas.csv](https://github.com/adempiere/adempiere/files/11423616/03h_Importar_Ordenes_de_Ventas.csv)

#### 3,000 Sales Orders consisting of one line per Sales Order: 
[03j_Importar_Ordenes_de_Ventas_una_linea_por_orden.csv](https://github.com/adempiere/adempiere/files/11423627/03j_Importar_Ordenes_de_Ventas_una_linea_por_orden.csv)

#### 1,000 Sales Orders consisting of one line per Sales Order:  
[03k_Importar_Ordenes_de_Ventas_una_linea_por_orden-1000_lineas.csv](https://github.com/adempiere/adempiere/files/11423633/03k_Importar_Ordenes_de_Ventas_una_linea_por_orden-1000_lineas.csv)

#### 2,000 Sales Orders consisting of one line per Sales Order:  
[03l_Importar_Ordenes_de_Ventas_una_linea_por_orden-2000_lineas.csv](https://github.com/adempiere/adempiere/files/11423638/03l_Importar_Ordenes_de_Ventas_una_linea_por_orden-2000_lineas.csv)

#### 1,000 Sales Orders consisting of one line per Sales Order:   
[03m_Importar_Ordenes_de_Ventas_una_linea_por_orden-1000_lineas.csv](https://github.com/adempiere/adempiere/files/11423639/03m_Importar_Ordenes_de_Ventas_una_linea_por_orden-1000_lineas.csv)

**Comment by e-Evolution** [comment]
@marcalwestf Hi Mario!

This is very useful for testing concurrency and high performance issues , in 3.9.4 I refactored the code to support parallel imports which makes import orders not sequential , we did some testing before releasing 3.9.4 I hope that in the environment that you are preparing the refactoring it will resist, in any case inform me of any strange behavior.

The important work is now to enable the same parallel import functionality for other processes, the work is hard and I will do it as clients demand it, but if anyone is interested in supporting this work I can explain how to refactor the code

**Comment by e-Evolution** [comment]
Hi @marcalwestf Mario , @piracio Horacio

If we talked about something over the weekend, about this topic, even though I understood that what he wants is to use the platform he contracted for one of his clients to do stress tests using ZK, but it implies creating scripts based on the platform to later run massive concurrency tests, I was testing the platform but it involved making proprietary scripts that won't be of much use after the license to use the platform runs out.

On the other hand, it happened to me that I could not make a test work correctly since each execution, the times between each step are variable and it had a strange behavior when executing again, something similar happened to me with JMeter, I hope that Horacio's experience and your experience help them to complete a successful test cycle.

It is imperative to find a solution to carry out more complex stress tests for adempiere, if we want ADempiere to be able to address other types of clients, since deficiencies must be detected and improvements implemented, only after finding the problems
